reducer.py

```diff
@@ -0,0 +1,319 @@
+
+from @armandmcqueen :
+
+> Can we come up with a more user-friendly approach than concatenation - dict
+of lists/2d array?
+
+(see the det.Reducer, which doesn't do any magic and gives access full access
+to raw metrics)
+
+from @aaron276h:
+
+> If we keep [@brain-good's strategy vs a hierarchichal reduce] configurable,
+even if there is a noticable performance hits shouldn't be a big a problem
+
+(det.SimpleReducer as a subclass of det.Reducer is designed to meet this need)
+
+
+(I was supposed to do some profiling work to measure the performance overhead
+of @brain-good's strategy vs a hierarchichal reduce strategy, but the truth is
+that I have not yet had time, and the more I thought about it the more I like
+@aaron276h's idea of making it configurable)
+"""
+
+
+class Reducer:
+    """
+    A two-stage hierarchical reduction.  For a simpler interface, see SimpleReducer, below.
+
+    Unlike the current internal reducers, both steps will always run, and users who want a simpler
+    interface will let the first step be an identity function.
+    """
+
+    @abc.abstractmethod
+    def per_slot_reduce(self, metrics):
+        """
+        d-train optimization: do the bulk of your reduction on each GPU.
+
+        metrics is literally just a list of the metrics your trial returned, no collation at all.
+
+        This is always called, in single- and multi-gpu settings.
+        """
+        pass
+
+    @abc.abstractmethod
+    def cross_slot_reduce(self, per_slot_metrics):
+        """
+        This is always called, in single- and multi-gpu settings.
+
+        per_slot_metrics will be the result of per_slot_reduce() on every worker, so
+        ``len(per_slot_metrics) == slots_per_trial``.
+        """
+        pass
+
+
+class MyAverageReducer(det.Reducer):
+    """Example implementation of a Reducer"""
+
+    def per_slot_reduce(self, metrics):
+        # Assume metrics is just a list of scalars
+        return [sum(metrics) / len(metrics), len(metrics)]
+
+    def cross_slot_reduce(self, per_slot_metrics):
+        """
+        Note that this step has to be a weighted average, even though the overall average is not a
+        weighted average.  This is the weird part of the hierarchical reduce.
+        """
+        weight_sum = sum(m[1] for m in per_slot_metrics)
+        return sum(val * weight for val, weight in per_slot_metrics) / weight_sum
+
+
+class MyWeightedAverageReducer(det.Reducer):
+    """Example implementation of a Reducer"""
+
+    def per_slot_reduce(self, metrics):
+        # Assume metrics is a list of [val, weight] for each batch
+        weight_sum = sum(m[1] for m in metrics)
+        return [sum(val * weight for val, weight in metrics), weight_sum]
+
+    def cross_slot_reduce(self, per_slot_metrics):
+        weight_sum = sum(m[1] for m in per_slot_metrics)
+        return sum(v * w for v, w in per_slot_metrics) / weight_sum
+
+#
+
+class SimpleReducer(det.Reducer):
```

Comment:
Having two types of user-facing reducers makes it a bit harder for users to understand (e.g. they need to know when to use which type and understand which type is being used in any given example). Do we think it is worth adding that surface area? I'm undecided - I prefer less surface area, but I do really like how clean the `MyAverageSimpleReducer` implementation becomes.